rust/src/storage/s3/mod.rs

```diff
@@ -315,34 +356,10 @@ impl StorageBackend for S3StorageBackend {
     }
 
     async fn rename_obj(&self, src: &str, dst: &str) -> Result<(), StorageError> {
-        debug!("rename s3 object: {} to {}...", src, dst);
-
-        match self.head_obj(dst).await {
-            Ok(_) => return Err(StorageError::AlreadyExists(dst.to_string())),
-            Err(StorageError::NotFound) => (),
-            Err(e) => return Err(e),
+        match self.lock_client {
+            None => self.unsafe_rename_obj(src, dst).await?,
```

Comment:
@houqp - @mosyp and I were discussing earlier, the combination of this bit and the silent default to `unsafe` makes us very nervous about folks running multiple workers accidentally corrupting the delta log by not providing an env var for `AWS_S3_LOCKING_PROVIDER`. 

I am thinking we should force configuration up front in some way for single or multi worker mode so we can fail fast if the locking provider is not configured in multi worker mode. Do you have any thoughts or preferences on a friendly/discoverable way to do this?

Reply:
We might set AWS_S3_LOCKING_PROVIDER to dynamodb by default, so if users want to use single-worker env they have to specify it explicitly

Reply:
Nice. I like the idea of defaulting the lock provider and requiring opt-in for single-worker mode. API-wise - I think an opt-in for single-worker mode might should come at a code level as an Option<> to the delta table factory functions. Relying on env vars would still make me feel nervous. Thinking about the difference between reader vs writer semantics is interesting though... I have no idea what the future of multi-reader looks like for delta-rs and whether this option would eventually be useful at a table level for readers.

Reply:
yep, +1 for explicit opt in for disabling dynamodb locking in s3 backend. another thing to consider is for read only users, we shouldn't force dynamodb on them. Perhaps by default, we would assuming dynamodb lock for s3, but if dynamodb is not configured, we would just throw an error in all write operations. That way, it still works transparently for read only consumers.

Reply:
yeah, makes sense, will roll out new changes to address your comments guys

Reply:
Another annoying thing for configuring backend is we don't really have an option to pass in storage backend specific option at the moment. The only thing that controls how backends are initialized is through uri scheme parsing, which is too limited. We will likely need to refactor that part of the code to make it more flexible.

For now, perhaps it's best to skip support for single writer configuration for now and move on with dynamodb only to unblock the project. We don't need single writer for our own use-cases anyway. Once we get stuff working end to end, we can come back to refactor the code to add single writer support.

Reply:
Single worker will not be possible right now, but I left it as is, e.g. users still need to explicitly configure it, which is ok, since. they have to specify dynamodb table at least. Also I like the idea of moving configs from envs into delta table options somehow, or into _main args_. But this is something that could be done later, once we have our mvp